### ISSUE TITLE
ZAProxy Plugin 2.0.0 Release Goals

### ISSUE THREAD

**Issue by JordanGS** [jenkinsci/zaproxy-plugin#42]
(**DONE**) bolded indicates completion of task or subset of task and has been added to #41 unless otherwise indicated.

1) **DONE** Custom Tools Plugin bug fixes and testing (bug fixed, requires wiki update)

2) **DEFER** JIRA Ticket Creation test (**[clone repo](https://github.com/JordanGS/security-tools)**, validate code, either deprecate or add disclaimer)

3) **DONE** Session Persist test (Session Load and Session Persist are now mutually exclusive, cannot choose both at same time)

4) **DONE** Figure out why selected format for report generation is not being saved on second click of configure.  (bug fixed)

5) **DONE** Host and Port are required in the job, that means that they overwrite the ones in admin controls, so what's the point? Investigate (bug fixed)

6) **DONE** Use parameters in methods and remove commented out code (Completed for Driver, other classes may require it as well)

7) **DONE** Update Wiki - @JordanGS

8) **DONE** Test run as Pre-Build

9) **DONE** Resolve/Close issues and pull requests on GitHub

Optional/Nice to Have but probably post 2.0.0 release

10) Use Jenkins credentials instead of fields

11) Build Management Tools - conditions for build failure/success

12) **DONE** UI Required Tag for text area? (textarea does not support required tag, added a hack text below the text area to inform the user that it's a required field unless there is a better way? Updated: added java end validation to the field instead)

13) Add HTTP/NTLM Authentication (Hostname, Port, Realm, Logged in Pattern)

14) **DONE** Add support for Export Report Plugin - @JordanGS

15) **DONE** Clearer Plugin messages

16) Execution of Arbitrary Script (Probably after the release of v2)
Before or After Authentication As an option. In case the script contains some crafted test cases for the target test, basically to contain a few tests in there.

17) **DONE** Upgrade the plugin to latest client 1.0 (Before the release of v2)

18) **DONE** Test all help icons, links to the wiki (Before the release of v2)

19) **DONE** Bug:Test pre-build/unexpected shut down should close the port (Before the release of v2)

20) **DONE** userAuthConfig in setUpUsers issue after upgrade to zap client 1.0 See issue (Before the release of v2) See Pull request #43

21) Fetch parameter names using the API instead of hard coding

22) Authenticated AJAX Spider

**Comment by JordanGS** [comment]
@mabdelmoez Completed are 1, 5, 9 and 12 (unless you have a suggestion or would like to look at it yourself). Currently working on 6, 7 and 11, 13 and 14 myself. If you could check at 3, 4 and 8 for selenium runs and 10 i would appreciate it or if you'd like some of the other tasks let me know. I have no issues either way :)

**Comment by mabdelmoez** [comment]
All good ! Lets merge that one.

**Comment by mabdelmoez** [comment]
@JordanGS I will add my goals later, they are very few since most of what i planned to add you have already taken care of. To be continued in slack.

**Comment by JordanGS** [comment]
@mabdelmoez all good i hope? I just did things are they came up down the workflow. We can discuss on slack.

**Comment by mabdelmoez** [comment]
**2-JIRA Ticket Creation:**
It might not be as important as it sounds, since opening JIRAs for Dynamic Analysis would end up opening tons of them (w/o very fine tuning and configuration), That's because Dynamic Analysis generates a lot of false positives, I think we should promote 14 (Add support for Export Report Plugin) instead, due to the fact that any automated runs/scans need reliable good reports right? Well ! If 2 is relatively easy to accomplish then we can do both in parallel 2 & 14. Given the Issues reported goes to JIRA.

**7-Update Wiki**
After you review and update the current, Lets agree that, whoever develops a feature or fixes a bug feeds it in the relevant documentation (help/notes/wiki), that would make the documentation go smoothly. Besides, who develops the feature or fixes the bug possesses the relevant knowledge for the documentation.

**8-Test run as Pre-Build** 
I always run ZAP in Pre-Build, so i know its issues and challenges.

**11-Build Management Tools**
Yes, that appeals to me, I had it on my list. I saw a JENKINS issue open for it tho.

**15-Clearer plugin messages**
As far as my runs went and the inconsistent messages I received, mostly It originates from ZAP API itself (especially 2.5), so lets discuss more on how to tackle that one.

I would add:

**16-Execution of Arbitrary Script** (Probably after the release of v2)
Before or After Authentication As an option. In case the script contains some crafted test cases for the target test, basically to contain a few tests in there. 

**17-Upgrade the plugin to use ZAP API 2.5** (Before the release of v2)

**18-Test all help icons.** (Before the release of v2)

Other things i wanted to add, is fixing a few bugs, but they disappeared since we actually changed a few things.

So I can take 8, 11, 16, 17, 18

After I test/push 8, 17, 18 will do a full test.

**Comment by JordanGS** [comment]
@mabdelmoez I'll review your pull tomorrow, it's late here now. But let's discuss on slack tomorrow around 11AM EST, that would be 5PM for you i believe? We can discuss next steps and everything. I'm fine with you taking 11, but the JIRA ticket you noticed was by my boss haha which was the main motivation behind them taking over the plugin. So let's discuss how we both envision that feature working and maybe we can get an ETA working on what's left so we have a timeline we're both comfortable with.

**Comment by JordanGS** [comment]
Completed task 15 (Updated all console output), also the following
- Fixed a bug with include in context
- Fixed a bug and console output for both form and script based authentication
- Fixed extra post data bug for form based authentication

**Comment by JordanGS** [comment]
-Finished 6 (Use parameters in methods and remove commented out code for ZAPDriver, leaving other classes alone)
-Fixed some additional bugs here and there.

**Comment by JordanGS** [comment]
#45 Merged Successfully

**Comment by mabdelmoez** [comment]
- Finished 16. Execution of Arbitrary Script (Probably after the release of v2) and created a pull request https://github.com/jenkinsci/zaproxy-plugin/pull/49

**Comment by JordanGS** [comment]
@mabdelmoez As per slack discussion 16) Execution of Arbitrary Script is completed but will be released as part of 2.1.0

**Comment by JordanGS** [comment]
2.0.0 Goalds completed, left as post 2.0.0 are

2) **DEFER** JIRA Ticket Creation test (**[clone repo](https://github.com/JordanGS/security-tools)**, validate code, either deprecate or add disclaimer)

11) Build Management Tools - conditions for build failure/success

13) Add HTTP/NTLM Authentication (Hostname, Port, Realm, Logged in Pattern)

16) Execution of Arbitrary Script (Probably after the release of v2)
Before or After Authentication As an option. In case the script contains some crafted test cases for the target test, basically to contain a few tests in there.

21) Fetch parameter names using the API instead of hard coding

22) Authenticated AJAX Spider
